File: ios/Data/Repositories/SearchRepository.swift
import Foundation

final class SearchRepository: SearchRepositoryProtocol {
    private let apiClient: APIClientProtocol

    init(apiClient: APIClientProtocol = APIClient.shared) {
        self.apiClient = apiClient
    }

    func search(query: String, type: SearchType?, cursor: String?, size: Int = 20) async -> RepositoryResult<UnifiedSearchResponse> {
        do {
            return .success(try await apiClient.request(SearchEndpoint.search(query: query, type: type, cursor: cursor, size: size)))
        } catch let error as APIError {
            return .failure(mapError(error))
        } catch {
            return .failure(.unknown)
        }
    }

    func searchRecipes(query: String, filters: RecipeFilters?, cursor: String?) async -> RepositoryResult<PaginatedResponse<RecipeSummary>> {
        let result = await search(query: query, type: .recipes, cursor: cursor)
        switch result {
        case .success(let response):
            let recipes = response.content.compactMap { item -> RecipeSummary? in
                if case .recipe(let recipe) = item.data {
                    return recipe
                }
                return nil
            }
            return .success(PaginatedResponse(
                content: recipes,
                nextCursor: response.nextCursor,
                hasNext: response.hasNext
            ))
        case .failure(let error):
            return .failure(error)
        }
    }

    func searchLogs(query: String, cursor: String?) async -> RepositoryResult<PaginatedResponse<CookingLogSummary>> {
        let result = await search(query: query, type: .logs, cursor: cursor)
        switch result {
        case .success(let response):
            let logs = response.content.compactMap { item -> CookingLogSummary? in
                if case .log(let log) = item.data {
                    return transformLogToSummary(log)
                }
                return nil
            }
            return .success(PaginatedResponse(
                content: logs,
                nextCursor: response.nextCursor,
                hasNext: response.hasNext
            ))
        case .failure(let error):
            return .failure(error)
        }
    }

    func searchUsers(query: String, cursor: String?) async -> RepositoryResult<PaginatedResponse<UserSummary>> {
        let result = await search(query: query, type: .users, cursor: cursor)
        switch result {
        case .success(let response):
            let users = response.content.compactMap { item -> UserSummary? in
                if case .user(let user) = item.data {
                    return user
                }
                return nil
            }
            return .success(PaginatedResponse(
                content: users,
                nextCursor: response.nextCursor,
                hasNext: response.hasNext
            ))
        case .failure(let error):
            return .failure(error)
        }
    }

    private func transformLogToSummary(_ log: LogPostSummaryResponse) -> CookingLogSummary {
        let recipe: RecipeSummary? = log.recipeTitle.map { title in
            RecipeSummary(
                id: "",
                title: title,
                description: nil,
                foodName: log.foodName ?? "",
                cookingStyle: nil,
                userName: "",
                thumbnail: nil,
                variantCount: 0,
                logCount: 0,
                servings: nil,
                cookingTimeRange: nil,
                hashtags: [],
                isPrivate: false
            )
        }
        return CookingLogSummary(
            id: log.id,
            rating: log.rating ?? 0,
            content: log.content,
            images: log.thumbnailUrl.map { [ImageInfo(id: log.id, url: $0, thumbnailUrl: $0, width: nil, height: nil)] } ?? [],
            author: UserSummary(id: log.creatorPublicId ?? "", username: log.userName, displayName: nil, avatarUrl: nil, level: 0, isFollowing: nil),
            recipe: recipe,
            likeCount: 0,
            commentCount: log.commentCount ?? 0,
            isLiked: false,
            isSaved: false,
            createdAt: Date()
        )
    }

    func getTrendingHashtags() async -> RepositoryResult<[HashtagCount]> {
        do {
            return .success(try await apiClient.request(SearchEndpoint.trending))
        } catch let error as APIError {
            return .failure(mapError(error))
        } catch {
            return .failure(.unknown)
        }
    }

    func getHashtagContent(hashtag: String, type: SearchType?, cursor: String?) async -> RepositoryResult<HashtagContentResponse> {
        do {
            return .success(try await apiClient.request(SearchEndpoint.hashtagContent(hashtag: hashtag, type: type, cursor: cursor)))
        } catch let error as APIError {
            return .failure(mapError(error))
        } catch {
            return .failure(.unknown)
        }
    }

    private func mapError(_ error: APIError) -> RepositoryError {
        switch error {
        case .networkError(let msg): return .networkError(msg)
        case .unauthorized: return .unauthorized
        case .notFound: return .notFound
        case .serverError(_, let msg): return .serverError(msg)
        case .decodingError(let msg): return .decodingError(msg)
        default: return .unknown
        }
    }
}

File: android/app/src/main/java/com/cookstemma/app/data/repository/SearchRepository.kt
package com.cookstemma.app.data.repository

import com.cookstemma.app.data.api.ApiService
import com.cookstemma.app.domain.model.*
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import javax.inject.Inject

// UI-friendly search results (transformed from API response)
data class SearchResults(
    val recipes: List<RecipeSummary> = emptyList(),
    val logs: List<FeedItem> = emptyList(),
    val users: List<UserSummary> = emptyList(),
    val hashtags: List<HashtagResult> = emptyList()
)

// API Response model matching backend
data class UnifiedSearchResponse(
    val content: List<SearchResultItem> = emptyList(),
    val counts: SearchCountsResponse? = null,
    val page: Int = 0,
    val size: Int = 20,
    val totalElements: Int = 0,
    val totalPages: Int = 0,
    val hasNext: Boolean = false,
    val nextCursor: String? = null
)

data class SearchCountsResponse(
    val recipes: Int = 0,
    val logs: Int = 0,
    val hashtags: Int = 0,
    val total: Int = 0
)

data class SearchResultItem(
    val type: String,
    val relevanceScore: Double? = null,
    val data: Map<String, Any?>? = null
)

data class HashtagResult(
    @SerializedName("publicId")
    val id: String = "",
    val name: String = "",
    @SerializedName("totalCount")
    val postCount: Int = 0
) {
    val tag: String get() = name
}

data class HashtagSearchResult(
    @SerializedName("publicId")
    val id: String = "",
    val name: String = "",
    val recipeCount: Int = 0,
    val logCount: Int = 0,
    val sampleThumbnails: List<String>? = null
) {
    val totalCount: Int get() = recipeCount + logCount
}

class SearchRepository @Inject constructor(
    private val apiService: ApiService
) {
    fun search(
        query: String,
        type: SearchType? = null,
        cursor: String? = null
    ): Flow<Result<SearchResults>> = flow {
        try {
            val response = apiService.search(query, type?.value, cursor)
            val results = transformSearchResponse(response)
            emit(Result.Success(results))
        } catch (e: Exception) {
            emit(Result.Error(e))
        }
    }

    private fun transformSearchResponse(response: UnifiedSearchResponse): SearchResults {
        val recipes = mutableListOf<RecipeSummary>()
        val logs = mutableListOf<FeedItem>()
        val hashtags = mutableListOf<HashtagResult>()

        response.content.forEach { item ->
            val data = item.data ?: return@forEach
            when (item.type) {
                "RECIPE" -> {
                    try {
                        val recipe = RecipeSummary(
                            id = (data["publicId"] as? String) ?: "",
                            title = (data["title"] as? String) ?: "",
                            description = data["description"] as? String,
                            foodName = (data["foodName"] as? String) ?: "",
                            cookingStyle = data["cookingStyle"] as? String,
                            userName = (data["userName"] as? String) ?: "",
                            thumbnail = data["thumbnail"] as? String,
                            variantCount = (data["variantCount"] as? Double)?.toInt() ?: 0,
                            logCount = (data["logCount"] as? Double)?.toInt() ?: 0,
                            servings = (data["servings"] as? Double)?.toInt(),
                            cookingTimeRange = data["cookingTimeRange"] as? String,
                            isPrivate = (data["isPrivate"] as? Boolean) ?: false
                        )
                        recipes.add(recipe)
                    } catch (e: Exception) { /* Skip malformed item */ }
                }
                "LOG" -> {
                    try {
                        val log = FeedItem(
                            id = (data["publicId"] as? String) ?: "",
                            title = data["title"] as? String,
                            content = data["content"] as? String,
                            rating = (data["rating"] as? Double)?.toInt(),
                            thumbnailUrl = data["thumbnailUrl"] as? String,
                            creatorPublicId = (data["creatorPublicId"] as? String) ?: "",
                            userName = (data["userName"] as? String) ?: "",
                            foodName = data["foodName"] as? String,
                            recipeTitle = data["recipeTitle"] as? String,
                            hashtags = (data["hashtags"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                            isVariant = data["isVariant"] as? Boolean,
                            isPrivate = data["isPrivate"] as? Boolean,
                            commentCount = (data["commentCount"] as? Double)?.toInt(),
                            cookingStyle = data["cookingStyle"] as? String
                        )
                        logs.add(log)
                    } catch (e: Exception) { /* Skip malformed item */ }
                }
                "HASHTAG" -> {
                    try {
                        val hashtag = HashtagResult(
                            id = (data["publicId"] as? String) ?: "",
                            name = (data["name"] as? String) ?: "",
                            postCount = ((data["recipeCount"] as? Double)?.toInt() ?: 0) +
                                       ((data["logCount"] as? Double)?.toInt() ?: 0)
                        )
                        hashtags.add(hashtag)
                    } catch (e: Exception) { /* Skip malformed item */ }
                }
            }
        }

        return SearchResults(
            recipes = recipes,
            logs = logs,
            users = emptyList(), // Users not included in unified search
            hashtags = hashtags
        )
    }

    fun searchRecipes(
        query: String,
        cursor: String? = null
    ): Flow<Result<PaginatedResponse<RecipeSummary>>> = flow {
        try {
            val response = apiService.search(query, SearchType.RECIPES.value, cursor)
            val recipes = extractRecipesFromResponse(response)
            emit(Result.Success(PaginatedResponse(
                content = recipes,
                nextCursor = response.nextCursor,
                hasMore = response.hasNext
            )))
        } catch (e: Exception) {
            emit(Result.Error(e))
        }
    }

    private fun extractRecipesFromResponse(response: UnifiedSearchResponse): List<RecipeSummary> {
        return response.content.mapNotNull { item ->
            if (item.type != "RECIPE") return@mapNotNull null
            val data = item.data ?: return@mapNotNull null
            try {
                RecipeSummary(
                    id = (data["publicId"] as? String) ?: "",
                    title = (data["title"] as? String) ?: "",
                    description = data["description"] as? String,
                    foodName = (data["foodName"] as? String) ?: "",
                    cookingStyle = data["cookingStyle"] as? String,
                    userName = (data["userName"] as? String) ?: "",
                    thumbnail = data["thumbnail"] as? String,
                    variantCount = (data["variantCount"] as? Double)?.toInt() ?: 0,
                    logCount = (data["logCount"] as? Double)?.toInt() ?: 0,
                    servings = (data["servings"] as? Double)?.toInt(),
                    cookingTimeRange = data["cookingTimeRange"] as? String,
                    isPrivate = (data["isPrivate"] as? Boolean) ?: false
                )
            } catch (e: Exception) {
                null
            }
        }
    }

    fun searchLogs(
        query: String,
        cursor: String? = null
    ): Flow<Result<PaginatedResponse<FeedItem>>> = flow {
        try {
            val response = apiService.search(query, SearchType.LOGS.value, cursor)
            val logs = extractLogsFromResponse(response)
            emit(Result.Success(PaginatedResponse(
                content = logs,
                nextCursor = response.nextCursor,
                hasMore = response.hasNext
            )))
        } catch (e: Exception) {
            emit(Result.Error(e))
        }
    }

    private fun extractLogsFromResponse(response: UnifiedSearchResponse): List<FeedItem> {
        return response.content.mapNotNull { item ->
            if (item.type != "LOG") return@mapNotNull null
            val data = item.data ?: return@mapNotNull null
            try {
                FeedItem(
                    id = (data["publicId"] as? String) ?: "",
                    title = data["title"] as? String,
                    content = data["content"] as? String,
                    rating = (data["rating"] as? Double)?.toInt(),
                    thumbnailUrl = data["thumbnailUrl"] as? String,
                    creatorPublicId = (data["creatorPublicId"] as? String) ?: "",
                    userName = (data["userName"] as? String) ?: "",
                    foodName = data["foodName"] as? String,
                    recipeTitle = data["recipeTitle"] as? String,
                    hashtags = (data["hashtags"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                    isVariant = data["isVariant"] as? Boolean,
                    isPrivate = data["isPrivate"] as? Boolean,
                    commentCount = (data["commentCount"] as? Double)?.toInt(),
                    cookingStyle = data["cookingStyle"] as? String
                )
            } catch (e: Exception) {
                null
            }
        }
    }

    fun searchUsers(
        query: String,
        cursor: String? = null
    ): Flow<Result<PaginatedResponse<UserSummary>>> = flow {
        try {
            val response = apiService.search(query, SearchType.USERS.value, cursor)
            val users = extractUsersFromResponse(response)
            emit(Result.Success(PaginatedResponse(
                content = users,
                nextCursor = response.nextCursor,
                hasMore = response.hasNext
            )))
        } catch (e: Exception) {
            emit(Result.Error(e))
        }
    }

    private fun extractUsersFromResponse(response: UnifiedSearchResponse): List<UserSummary> {
        return response.content.mapNotNull { item ->
            if (item.type != "USER") return@mapNotNull null
            val data = item.data ?: return@mapNotNull null
            try {
                UserSummary(
                    id = (data["publicId"] as? String) ?: "",
                    username = data["username"] as? String,
                    displayName = data["displayName"] as? String,
                    avatarUrl = data["profileImageUrl"] as? String
                )
            } catch (e: Exception) {
                null
            }
        }
    }

    fun getTrendingHashtags(): Flow<Result<List<HashtagResult>>> = flow {
        try {
            val response = apiService.getTrendingHashtags()
            emit(Result.Success(response))
        } catch (e: Exception) {
            emit(Result.Error(e))
        }
    }

    fun getHashtagPosts(
        hashtag: String,
        cursor: String? = null
    ): Flow<Result<PaginatedResponse<FeedItem>>> = flow {
        try {
            val response = apiService.getHashtagPosts(hashtag, cursor)
            emit(Result.Success(response))
        } catch (e: Exception) {
            emit(Result.Error(e))
        }
    }
}

enum class SearchType(val value: String) {
    RECIPES("recipes"),
    LOGS("logs"),
    USERS("users"),
    HASHTAGS("hashtags")
}
